src/cpp/shared_core/system/encryption/EncryptionVersionTests.cpp

```diff
@@ -109,90 +116,181 @@ test_context("EncryptionVersionTests")
       REQUIRE_FALSE(error);
+
+      // decrypt the encrypted data
       std::vector<unsigned char> decryptedData;
-      error = core::system::crypto::v1::aesDecrypt(encryptedData, g_key, g_iv, decryptedData);
+      error = core::system::crypto::v2::aesDecrypt(encryptedData, g_key, g_iv, decryptedData);
       REQUIRE_FALSE(error);
+
+      // verify that the decryption gives us back the original data
       REQUIRE(decryptedPayloadMatches(decryptedData));
+   }
 
-      // re-setup
-      REQUIRE(generateKeys());
-      encryptedData = {};
-      decryptedData = {};
+   test_that("v2: Can AES encrypt/decrypt detect tampering/corruption")
+   {
+      // setup
+      REQUIRE(generateKeys(crypto::v2::VERSION_BYTE));
 
-      // encrypt with v1, decrypt with v0
-      error = core::system::crypto::v1::aesEncrypt(g_data, g_key, g_iv, encryptedData);
+      // encrypt the data
+      std::vector<unsigned char> encryptedData;
+      Error error = core::system::crypto::v2::aesEncrypt(g_data, g_key, g_iv, encryptedData);
       REQUIRE_FALSE(error);
-      error = core::system::crypto::v0::aesDecrypt(encryptedData, g_key, g_iv, decryptedData);
+
+      // decrypt the encrypted data with wrong AAD
+      std::vector<unsigned char> decryptedData;
+      std::vector<unsigned char> bad_aad = encryptedData;
+      bad_aad[crypto::VERSION_BYTE_INDEX] += 1;
+      error = core::system::crypto::v2::aesDecrypt(bad_aad, g_key, g_iv, decryptedData);
+      REQUIRE(decryptedData.size() == 0);
+      REQUIRE(error != Success());
+
+      // decrypt the encrypted data with changed data
+      std::vector<unsigned char> bad_encryptedData = encryptedData;
+      bad_encryptedData[crypto::ENCRYPTION_VERSION_SIZE_BYTES] += 1;
+      error = core::system::crypto::v2::aesDecrypt(bad_encryptedData, g_key, g_iv, decryptedData);
+      REQUIRE(decryptedData.size() == 0);
+      REQUIRE(error != Success());
+
+      // decrypt the encrypted data with wrong MAC
+      std::vector<unsigned char> bad_mac = encryptedData;
+      bad_mac.back() += 1;
+      error = core::system::crypto::v2::aesDecrypt(bad_mac, g_key, g_iv, decryptedData);
+      REQUIRE(decryptedData.size() == 0);
+      REQUIRE(error != Success());
+
+      // Finally, decrypt the encrypted data correctly
+      error = core::system::crypto::v2::aesDecrypt(encryptedData, g_key, g_iv, decryptedData);
+      REQUIRE(decryptedData.size() > 0);
       REQUIRE_FALSE(error);
+
+      // verify that the decryption gives us back the original data
       REQUIRE(decryptedPayloadMatches(decryptedData));
    }
 
-   test_that("v2: Can AES encrypt/decrypt")
+   test_that("Key size mismatches aren't allowed")
+   {
+      std::vector<unsigned char> iv(16);
+      std::vector<unsigned char> encryptedData;
+
+      // Encrypting v0/v1 with v2 key size throws
+      std::vector<unsigned char> key(crypto::v2::KEY_LENGTH_BYTES);
+      REQUIRE_THROWS(core::system::crypto::v0::aesEncrypt(g_data, key, iv, encryptedData));
```

Comment:
Looks like this isn't throwing an exception so the test is failing in the PR build.